File: modules/protocol/base/src/main/java/candle/protocol/ConnectionState.java
package candle.protocol;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

import java.util.ArrayList;

@Getter
@EqualsAndHashCode
@ToString
public class ConnectionState {
  private static final ArrayList<ConnectionState> STATES = new ArrayList<>();

  public static final ConnectionState HANDSHAKING = ConnectionState.create("Handshaking");
  public static final ConnectionState CONFIGURATION = ConnectionState.create("Configuration");
  public static final ConnectionState LOGIN = ConnectionState.create("Login");
  public static final ConnectionState PLAY = ConnectionState.create("Play");
  public static final ConnectionState STATUS = ConnectionState.create("Status");

  private final String name;

  ConnectionState(String name) {
    this.name = name;
  }

  public static ConnectionState create( String name ) {
    ConnectionState state = new ConnectionState(name);
    STATES.add(state);
    return state;
  }
}

File: protocol/base/src/main/kotlin/net/candlemc/protocol/ConnectionState.kt
package net.candlemc.protocol

class ConnectionState(private val name: String) {
    override fun toString(): String {
        return name
    }

    companion object {
        @JvmStatic
        val HANDSHAKING = ConnectionState("Handshaking")
        @JvmStatic
        val STATUS = ConnectionState("Status")
        @JvmStatic
        val LOGIN = ConnectionState("Login")
        @JvmStatic
        val CONFIGURATION = ConnectionState("Configuration")
        @JvmStatic
        val PLAY = ConnectionState("Play")
        @JvmStatic
        val TRANSFER = ConnectionState("Transfer")
    }
}
